golem-cli/src/cloud/auth.rs

```diff
@@ -0,0 +1,196 @@
+// Copyright 2024 Golem Cloud
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fs::{create_dir_all, File, OpenOptions};
+use std::io::{BufReader, BufWriter};
+use std::path::{Path, PathBuf};
+
+use async_trait::async_trait;
+use chrono::{DateTime, Utc};
+use golem_cloud_client::model::{OAuth2Data, Token, TokenSecret, UnsafeToken};
+use indoc::printdoc;
+use serde::{Deserialize, Serialize};
+use tracing::info;
+use uuid::Uuid;
+
+use crate::cloud::clients::login::LoginClient;
+use crate::cloud::clients::CloudAuthentication;
+use crate::model::GolemError;
+
+#[async_trait]
+pub trait Auth {
+    async fn authenticate(
+        &self,
+        manual_token: Option<Uuid>,
+        config_dir: PathBuf,
+    ) -> Result<CloudAuthentication, GolemError>;
+}
+
+pub struct AuthLive {
+    pub login: Box<dyn LoginClient + Send + Sync>,
+}
+
+#[derive(Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct CloudAuthenticationConfig {
+    data: CloudAuthenticationConfigData,
+    secret: Uuid,
+}
+
+#[derive(Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct CloudAuthenticationConfigData {
+    id: Uuid,
+    account_id: String,
+    created_at: DateTime<Utc>,
+    expires_at: DateTime<Utc>,
```

Comment:
Didn't know this much about cloud is going to get exposed into OSS in this work.

Reply:
Let's discuss it in slack